From b726a845433596e9b169312f280df1aa15e52fc6 Mon Sep 17 00:00:00 2001
From: Nobuhiko Tanibata <NOBUHIKO_TANIBATA@xddp.denso.co.jp>
Date: Tue, 26 Aug 2014 08:48:06 -0700
Subject: [PATCH 22/31] input.c: give control of default_grab_keyboard_key to a
 shell.

In automotive usecase, a surface which is not focused also wants to
receive keyboard event. E.g. Media Player in background wants to receive
volume up/down.

Signed-off-by: Nobuhiko Tanibata <NOBUHIKO_TANIBATA@xddp.denso.co.jp>
---
 src/input.c | 29 ++++++++++++++++++++---------
 1 file changed, 20 insertions(+), 9 deletions(-)

diff --git a/src/input.c b/src/input.c
index 2c799f4..8e0dac3 100644
--- a/src/input.c
+++ b/src/input.c
@@ -34,6 +34,13 @@
 #include "../shared/os-compatibility.h"
 #include "compositor.h"
 
+// for input focus step1
+typedef void
+(*keyboard_key)(struct weston_keyboard_grab *grab,
+                uint32_t time, uint32_t key, uint32_t state);
+
+WL_EXPORT keyboard_key keyboard_key_func=NULL;
+
 static void
 empty_region(pixman_region32_t *region)
 {
@@ -318,15 +325,19 @@ default_grab_keyboard_key(struct weston_keyboard_grab *grab,
 	uint32_t serial;
 	struct wl_list *resource_list;
 
-	resource_list = &keyboard->focus_resource_list;
-	if (!wl_list_empty(resource_list)) {
-		serial = wl_display_next_serial(display);
-		wl_resource_for_each(resource, resource_list)
-			wl_keyboard_send_key(resource,
-					     serial,
-					     time,
-					     key,
-					     state);
+	if (keyboard_key_func) {
+		keyboard_key_func(grab, time, key, state);
+	} else {
+		resource_list = &keyboard->focus_resource_list;
+		if (!wl_list_empty(resource_list)) {
+			serial = wl_display_next_serial(display);
+			wl_resource_for_each(resource, resource_list)
+				wl_keyboard_send_key(resource,
+						     serial,
+						     time,
+						     key,
+						     state);
+		}
 	}
 }
 
-- 
1.9.3

